### ISSUE TITLE
Warning: Multiple instances of Vue detected!

### ISSUE THREAD

**Issue by groboter** [bootstrap-vue/bootstrap-vue#3351]
I'm using electron-vue and bootstrap-vue and getting the warning "Multiple instances of Vue detected!".

Whole warning message:
```
D:\temp\test1\node_m…ap-vue.common.js:80 [BootstrapVue warn]: Multiple instances of Vue detected!
See: https://bootstrap-vue.js.org/docs#using-module-bundlers
warn	@	D:\temp\test1\node_m…ap-vue.common.js:80
checkMultipleVue	@	D:\temp\test1\node_m…-vue.common.js:1413
install	@	D:\temp\test1\node_m…-vue.common.js:1437
Vue.use	@	vue.esm.js?a026:5110
install	@	D:\temp\test1\node_m…vue.common.js:23791
Vue.use	@	vue.esm.js?a026:5110
(anonymous)	@	main.js?cebd:9
./src/renderer/main.js	@	renderer.js:2652
__webpack_require__	@	renderer.js:725
fn	@	renderer.js:102
0	@	renderer.js:2723
__webpack_require__	@	renderer.js:725
(anonymous)	@	renderer.js:792
(anonymous)	@	renderer.js:795
```
I'm a js beginner and for me it is not obvious where the other vue instance(s) are generated and how I can remove or suppress them?

**Comment by tmorehouse** [comment]
See the getting started docs section on bundlers: https://bootstrap-vue.js.org/docs#using-module-bundlers (specifically the section at the bottom of the section regarding aliases.

**Comment by groboter** [comment]
I have already installed the package according to the mentioned docs section using `npm i` and also registered the plugin in my app entry point. My `webpack.renderer.config.js` contains the following entry:
```
  resolve: {
    alias: {
      '@': path.join(__dirname, '../src/renderer'),
      'vue$': 'vue/dist/vue.esm.js'
    },
    extensions: ['.js', '.vue', '.json', '.css', '.node']
  },
```

**Comment by SergioFloresG** [comment]
Hello, same problem, and probe everything indicated with webpack and sige showing the alert.

## main.js
```js
import Vue from 'vue'
import axios from 'axios'

//
// Bootstrap 4
//
const BootstrapVue = require('bootstrap-vue');
Vue.use(BootstrapVue);

//
// Main App
//
import './main.scss';
import App from './App'

Vue.http = Vue.prototype.$http = axios;
Vue.config.productionTip = false;

new Vue({
  components: {App},
  render(h) {
    return h(App)
  }
}).$mount('#app');
```

## main.scss
```scss
@import "~bootstrap/scss/bootstrap.scss";
@import "~bootstrap-vue/src/index";
```

## webpack.config.js
```js
module.exports = {
   resolve: {
    alias: {
      '@': path.join(__dirname, '../src/renderer'),
      'vue$': 'vue/dist/vue.esm.js'
    }
  },
}
```

**Comment by tmorehouse** [comment]
Try changing this:

```js
const BootstrapVue = require('bootstrap-vue');
Vue.use(BootstrapVue);
```

To this
```js
import BootstrapVue from 'bootstrap-vue';

Vue.use(BootstrapVue);
```

**Comment by SergioFloresG** [comment]
Mmm... For some reason if i use import the library is not carge.

I put this and magic happens, no warning appears and works perfectly.

## main.js
```js
Vue.use(require('bootstrap-vue/dist/bootstrap-vue.common.min'));
```

## webpack.config.js
But i'm confused in mi _webpack.config.js_ the configuration is this:
```js
module.exports = {
  output: {
    filename: '[name].js',
    libraryTarget: 'commonjs2',
    path: path.join(__dirname, '../dist/electron')
  },
  resolve: {
    alias: {
      '@': path.join(__dirname, '../src/renderer'),
      'vue$': 'vue/dist/vue.esm.js'
    },
    extensions: ['.js', '.vue', '.json', '.css', '.node']
  }
}
```

So, \ ( * .* ) magic.

**Comment by tmorehouse** [comment]
Hmm, When `require` is placed inside `Vue.use(...)` it is lazy loading (via WebPack) BootstrapVue when `Vue.use(...)` is executed.

Are you using Webpack 4?

**Comment by groboter** [comment]
For me the magic `Vue.use(require('bootstrap-vue/dist/bootstrap-vue.common.min'));` also works.

**Comment by groboter** [comment]
I'm using webpack `webpack@4.32.2`

**Comment by SergioFloresG** [comment]
> I'm using webpack `webpack@4.32.2`

For me: webpack@4.15.1

**Comment by tmorehouse** [comment]
Cross reference issue with Vue: https://github.com/vuejs/vue/issues/8278

**Comment by tmorehouse** [comment]
@SergioFloresG I noticed this comment in another library comment regarding `electron` https://github.com/vuetifyjs/vuetify/issues/4068#issuecomment-491091793:

> Search electron project files for the variable "whiteListedModules", I found it in .electron-vue/webpack.renderer.config.js
> 
> let whiteListedModules = ['vue']
>
> In my scenario, I had added bootstrap-vue package to my project and whenever I was interacting with vue components supplied from bootstrap-vue I was encountering the error, this means that the bootstrap-vue was loading in the second instance of Vue which causes the $attr readonly errors and similar errors.
>
> In order prevent you application having two instances of Vue, you need to add the offending package into the whiteListedModules array, along with vue:
>
> let whiteListedModules = ['vue', 'bootstrap-vue']
> 
> Now I do not get the $attr readonly errors or anything else, and vue and bootstrap vue run fine.

**Comment by tmorehouse** [comment]
If you have multiple webpack configs (i.e. `webpack.config.js`, `webpack.renderer.config.js`, `webpack.vendor.config.js`, etc), you will need to set the alias in all of them

**Comment by groboter** [comment]
Works perfect. Thanks @tmorehouse!

**Comment by mudlee** [comment]
The interesting this is that this error only happens if I use the code with electron-webpack. I use bootstrap-vue without any problem in my vue project, but as soon as I put the code under electron-webpack, I get this error.

If I switch back the version to 2.0.0-rc.15, then the error disappears.

However, in both cases with electron-webpack the styles somewhy do not get applied. Again, in a simple vue project, everything is fine. Under electron-webpack, I get the error with the latest version AND I got not styling.

I have a webpack config in package.json:
```json
...
"electronWebpack": {
    "main": {
      "webpackConfig": "webpack.config.js"
    }
}
```

and in the webpack.config.js:
```json
module.exports = {
  "resolve": {
    "alias": {
      "vue": "vue/dist/vue.runtime.esm.js" // yes, I need the runtime
    }
  }
}
```

Any idea?

**Comment by loopmode** [comment]
Please see https://github.com/bootstrap-vue/bootstrap-vue/issues/3623#issuecomment-508908862 and effectively https://github.com/electron-userland/electron-webpack/issues/302#issuecomment-508911966

1) As stated in https://github.com/bootstrap-vue/bootstrap-vue/issues/3351#issuecomment-495938259 by @tmorehouse , whitelisting the frontend libraries removes the warning about duplicate versions. In an `electron-webpack` context, the alias has nothing to do with the warnings and it's not necessary to provide a custom webpack config, as `electron-webpack-vue` already provides the aliases.

See https://github.com/loopmode/electron-webpack-quick-start/tree/bootstrap-vue-ts and it's two last commits for a working example.

**Comment by rohitrameshrao** [comment]
> @SergioFloresG @groboter I noticed this comment in another library comment regarding `electron` [vuetifyjs/vuetify#4068 (comment)](https://github.com/vuetifyjs/vuetify/issues/4068#issuecomment-491091793) by @DanWithams:
> 
> > Search electron project files for the variable "whiteListedModules", I found it in .electron-vue/webpack.renderer.config.js
> > let whiteListedModules = ['vue']
> > In my scenario, I had added bootstrap-vue package to my project and whenever I was interacting with vue components supplied from bootstrap-vue I was encountering the error, this means that the bootstrap-vue was loading in the second instance of Vue which causes the $attr readonly errors and similar errors.
> > In order prevent you application having two instances of Vue, you need to add the offending package into the whiteListedModules array, along with vue:
> > let whiteListedModules = ['vue', 'bootstrap-vue']
> > Now I do not get the $attr readonly errors or anything else, and vue and bootstrap vue run fine.

This is missing from the documentation page. Currently, the doc talks just about alias and not about white-listing.

**Comment by loopmode** [comment]
I believe whiteListedModules, especially with that casing, is specific to `electron-webpack`. However, the key takeaway is this: when working with webpack and electron, typically all dependencies are defined as `externals` in webpack config. While this is rather necessary for modules used by the main process (it needs to resolve from node_modules at runtime), this is unnecessary and sometimes problematic for modules used by the renderer process, which should have all its required modules bundled at compile time.

So, long story short: when working with `electron-webpack`, use the `whiteListedModules` to exclude particular frontend packages from webpack externals.
When using plain webpack with electron, do not add those packages to `externals| config in the first place.

**Comment by Innerfree** [comment]
> For me the magic `Vue.use(require('bootstrap-vue/dist/bootstrap-vue.common.min'));` also works.
That's a magic ,it works for me too.

**Comment by imanau** [comment]
Hellow!
I founded same ploblem.

I use browserify and gulp in my project

Then, I change my `package.json`  to setting alias of 'vue'

```

"browser": {
    "vue": "vue/dist/vue.common.js"
  }

```

But  the error message does not disappear
`[BootstrapVue warn]: Multple instance of Vue detected!` 

I tried that method too, but it didn't work
> For me the magic Vue.use(require('bootstrap-vue/dist/bootstrap-vue.common.min')); also works.
That's a magic ,it works for me too.

![image](https://user-images.githubusercontent.com/46194836/83365595-2d884e00-a3e4-11ea-8315-67eb5887da62.png)


Any idea?

**Comment by deni5n** [comment]
I have the same problem with Nuxtjs. I am doing an integration test and I see a warning:
    [BootstrapVue warn]: Multiple instances of Vue detected!
    You may need to set up an alias for Vue in your bundler config.
    See: https://bootstrap-vue.org/docs#using-module-bundlers

My nuxt.config.js has build section:
```
build: {
    extend(config, ctx) {
      config.resolve.alias.vue$ = path.resolve( __dirname, 'node_modules/vue/dist/vue.runtime.esm.js')
    },
},
```
my test file:
```
/**
 * @jest-environment node
 */
const resolve = require('path').resolve
const { Nuxt, Builder } = require('nuxt')
let nuxt = null

beforeAll(async () => {
  const rootDir = resolve(__dirname, '../')
  let config = {}
  config = require(resolve(rootDir, 'nuxt.config.js')).default
  nuxt = new Nuxt(config)
  await new Builder(nuxt).build()
  nuxt.listen(3001, 'localhost')
}, 20000)

afterAll(() => {
  nuxt.close()
})

describe('example', () => {
  it('stub', () => {
    expect(1).toBe(1)
  })
})
```
How I can remove duplicate instance?
